src/libraries/System.Private.Uri/src/System/Uri.cs

```diff
@@ -4650,77 +4593,84 @@ private unsafe char[] GetCanonicalPath(char[] dest, ref int pos, UriFormat forma
             //
             // (path is already  >= 3 chars if recognized as a DOS-like)
             //
-            if (dosPathIdx != 0 && dest[dosPathIdx + pos - 1] == '|')
-                dest[dosPathIdx + pos - 1] = ':';
+            if (dosPathIdx != 0 && dest[dosPathIdx + start - 1] == '|')
+                dest[dosPathIdx + start - 1] = ':';
 
             if (InFact(Flags.ShouldBeCompressed))
             {
                 // It will also convert back slashes if needed
-                dest = Compress(dest, (ushort)(pos + dosPathIdx), ref end, _syntax);
-                if (dest[pos] == '\\')
-                    dest[pos] = '/';
+                dest.Length = start + Compress(dest.RawChars.Slice(start, dest.Length - start), _syntax);
+                if (dest[start] == '\\')
+                    dest[start] = '/';
 
                 // Escape path if requested and found as not fully escaped
                 if (formatAs == UriFormat.UriEscaped && NotAny(Flags.UserEscaped) && InFact(Flags.E_PathNotCanonical))
                 {
                     //Note: Flags.UserEscaped check is solely based on trusting the user
-                    dest = UriHelper.EscapeString(new string(dest, pos, end - pos), dest, ref pos, checkExistingEscaped: !IsImplicitFile, '?', '#');
-                    end = pos;
+
+                    // We can't do an in-place escape, create a copy
+                    ValueStringBuilder copy = new ValueStringBuilder(stackalloc char[StackallocThreshold]);
+                    copy.Append(dest.AsSpan(start, dest.Length - start));
+
+                    dest.Length = start;
+                    ReadOnlySpan<char> copySpan = MemoryMarshal.CreateReadOnlySpan(ref copy.GetPinnableReference(), copy.Length);
```

Comment:
Same as https://github.com/dotnet/runtime/pull/34864#discussion_r407184861 for using this instead of `copy.AsSpan()`